### ISSUE TITLE
Thread Tagging

### ISSUE THREAD

**Issue by mxstbr** [withspectrum/spectrum#4405]
We should add an optional thread tagging system similar to GitHub labels.

**High level description**

A whitelist of tags should be able to be defined on a community-level, which can then be used by moderators and owners to tag individual threads.

** Technical sketch**

- Store tags in a `communitiesThreadTags` table, with records that look like this:

  ```JS
  {
    id: 'fdgyit-123',
    title: 'some-string',
    communityId: 'dfqg-community-234'
  }
  ```

**Comment by superbryntendo** [comment]
re: user tagging, we could allow them to tag it initially, but if a moderator overrides it, we lock it.

we've definitely talked about tags since the beginning, but i'm not super sure of the use-case beyond what channels already does. could you add some examples, please?

**Comment by mxstbr** [comment]
For example:

1. Marking a thread as solved (open, closed)
1. Marking a thread team member replied
1. Assigning a team member or community member to a thread

**Comment by superbryntendo** [comment]
Should these be manual tags? or should they be automated (e.g. make a Q&A thread type that allows the creator to mark it as solved, automatically tag the thread on team member initial reply, etc.)

**Comment by mxstbr** [comment]
Manual tags I think, just let them manage their community however they want for the most flexibility!

**Comment by brianlovin** [comment]
> Can users tag their own threads?

I don't think so - maybe to clarify we should consider this a team/moderator feature, and not even show it to users? I think that channels are probably the appropriate way for community members to get their content in the right context.

This removes the problem of having to have tags search-indexed.

Question: should we denormalize the tags on the thread object so that we can add a secondary index to the `threads` table? This would allow us to have a sort option on the frontend for community owners to get threads by tag type. The only problem with denormalization would be that if a community updated/deleted an existing tag, we'd have to find all threads that use that tag and update it accordingly. That'd be a pain :/

From the user experience point of view, I think it's important that community team members be able to view threads of a given tag. So let's figure that one out here, too :)

**Comment by brianlovin** [comment]
> and not even show it to users?

If you disagree with this, then maybe the model for a `communityThreadTag` becomes:
```js
{
  id: string,
  communityId: string,
  label: string,
  color: string,
  isPublic: boolean,
}
```

**Comment by mxstbr** [comment]
We don't even have to denormalize theoretically, could just do something like:

```JS
// communitiesThreadTags
{
  id: ID,
  communityId: string,
  title: string,
  color: string,
}

// threads
{
  ...,
  tags: [ID]
}
```

No need for a tags/threads join table and it should make it possible to filter tags from the frontend fairly performantly!

Let's keep all tags public for now, adding `isPublic` down the line would be a relatively easy change if folks ask for it.

**Comment by brianlovin** [comment]
> Let's keep all tags public, adding isPublic down the line would be a relatively easy change if folks ask for it

Sure

> Let's not have colors, those would add a lot of complexity UI-wise. If everybody starts asking for colors, again it'd be a fairly easy change down the line

We can start without, but I'm sure people will ask for this right away since colors are such a useful mechanism for scanning and differentiating pieces of content.

**Comment by superbryntendo** [comment]
there's one use case that would be inconvenient for public tagging imo and that's assigning teammates to a thread.

I don't have strong feelings here, but I think this makes them less useful for that tag.

**Comment by mxstbr** [comment]
Hmm how's that inconvenient? It's the same on GitHub, right? Assignees are public

**Comment by superbryntendo** [comment]
support mechanisms are different than engineering though.

imagine if every time we assigned something to you in HelpScout, the user saw it and then was frustrated if you didn't answer it quickly.
